lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -4729,6 +4729,127 @@ struct PrintfOpConversion
+                  ConversionPatternRewriter &rewriter) const override {
+
+    auto loc = op.getLoc();
+    MLIRContext *ctx = rewriter.getContext();
+
+    auto atomicRmwAttr = op.atomic_rmw_op();
+    Value ptr = op.ptr();
+    Value val = op.val();
+
+    auto llAtomicRmwAttr = adaptor.atomic_rmw_op();
+    Value llPtr = adaptor.ptr();
+    Value llVal = adaptor.val();
+
+    auto valElements = getElementsFromStruct(loc, llVal, rewriter);
+    auto ptrElements = getElementsFromStruct(loc, llPtr, rewriter);
+
+    // TODO: Support mask and scalar
+
+    auto valueTy = op.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!valueTy)
+      return failure();
+    Type valueElemTy =
+        getTypeConverter()->convertType(valueTy.getElementType());
+
+    auto valTy = val.getType().cast<RankedTensorType>();
+    const size_t valueElemNbits = valueElemTy.getIntOrFloatBitWidth();
+    auto vec = getVectorSize(ptr);
+    vec = std::min<unsigned>(vec, valTy.getElementType().isF16() ? 2 : 1);
+
+    auto vecTy = LLVM::getFixedVectorType(valueElemTy, vec);
+    auto elemsPerThread = getElemsPerThread(val.getType());
+    SmallVector<Value> resultVals(elemsPerThread);
+    for (size_t i = 0; i < elemsPerThread; i += vec) {
+      Value rmvVal = undef(vecTy);
+      for (int ii = 0; ii < vec; ii++) {
+        Value iiVal = createIndexAttrConstant(
+            rewriter, loc, getTypeConverter()->getIndexType(), ii);
+        rmvVal = insert_element(vecTy, rmvVal, valElements[i], iiVal);
+      }
+      Value rmwPtr = bitcast(ptrElements[i], ptr_ty(valTy.getElementType()));
+      std::string s_ty{""};
+      PTXBuilder ptxBuilder;
+      auto &atom = *ptxBuilder.create<>("atom");
+      atom.o("global").o("gpu");
+      auto rmwOp = stringifyRMWOp(atomicRmwAttr).str();
+      auto sBits = std::to_string(valueElemNbits);
+      switch (atomicRmwAttr) {
+      case RMWOp::AND:
+        s_ty = "b" + sBits;
+        break;
+      case RMWOp::OR:
+        s_ty = "b" + sBits;
+        break;
+      case RMWOp::XOR:
+        s_ty = "b" + sBits;
+        break;
+      case RMWOp::ADD:
+        s_ty = "s" + sBits;
+        break;
+      case RMWOp::FADD:
+        rmwOp = "add";
+        rmwOp += (valueElemNbits == 16 ? ".noftz" : "");
+        s_ty = "f" + sBits;
+        s_ty += (vec == 2 && valueElemNbits == 16) ? "x2" : "";
+        break;
+      case RMWOp::MAX:
+        s_ty = "s" + sBits;
+        break;
+      case RMWOp::MIN:
+        s_ty = "s" + sBits;
+        break;
+      case RMWOp::UMAX:
+        rmwOp = "max";
+        s_ty = "u" + sBits;
+        break;
+      case RMWOp::UMIN:
+        rmwOp = "min";
+        s_ty = "u" + sBits;
+        break;
+      default:
+        return failure();
+      }
+      atom.o(rmwOp).o(s_ty);
+
+      auto *dstOpr = ptxBuilder.newOperand("=r");
+      auto *ptrOpr = ptxBuilder.newAddrOperand(rmwPtr, "r");
+      auto *valOpr = ptxBuilder.newOperand(rmvVal, "r");
+
+      atom(dstOpr, ptrOpr, valOpr);
```

Comment:
Move `ptxBuilder.create` here, and make the atom-create-related code more compat seems better, just personal teste.
The common procedure to use ptxBuilder is like, first define the operands, then create the ptxBuilder instruction, third, call the instruction with its operands, finally launch the ptx and process the result.

Reply:
Change the calling order of the PTXBuilder to define, create, call and launch. But keep the modifier gen code.